### PR TITLE
Change observer structure to contain State.

### PR DESCRIPTION
**Pull request by jessearens**
With State, observers can adjust their updating to only update when a
change happens to the certain state. Therefore the observer for f.e.
EnemyKills will not be updated when something else changes in the
PlayerFish class.

<feature,organisation

### CODE REVIEW

**Comment by jessearens** [comment]
Only one commit because I want to continue on the achievements from another branch. 

<review

File: src/main/java/com/github/fishio/PlayerFish.java [hunk 1]
```diff
@@ -266,7 +268,8 @@ public void preMove() {
 	}
 
 	@Override
-	public void hitWall() { }
+	public void hitWall() {
```
Comment (ChielBruin) [inline src/main/java/com/github/fishio/PlayerFish.java:273]: :P

File: src/main/java/com/github/fishio/PlayerFish.java [hunk 2]
```diff
@@ -292,14 +295,18 @@ public void onCollide(ICollidable other) {
 			double osize = fish.getBoundingArea().getSize();
 
 			if (tsize > osize * FISH_EAT_THRESHOLD) {
+				State old = getState();
+				
 				fish.setDead();
-				notifyObservers();
+				notifyObservers(old, getState());
```
Comment (Taeir) [inline src/main/java/com/github/fishio/PlayerFish.java]: do you mean notifyObservers(old, getState(), "death");

<bug

**Comment by Taeir** [comment]
I added one comment. Once that is resolved, its good to be merged.

File: src/main/java/com/github/fishio/achievements/Subject.java [hunk 3]
```diff
@@ -19,11 +26,23 @@
...
+	default void detach(Observer observer) {
+		getObservers().remove(observer);
+	}
 	
 	/**
 	 * Gives all the attached Observers a notification when something changes.
 	 * 
 	 */
-	void notifyObservers();
+	default void notifyObservers(State oldState, State newState, String... changed) {
```
Comment (ChielBruin) [inline src/main/java/com/github/fishio/achievements/Subject.java:47]: Javadoc missing
<documentation

**Comment by jessearens** [comment]
Resolved comments. Merged master into branch to fix the random failing builds which were already fixed in master.
<bug,review

File: src/main/java/com/github/fishio/achievements/State.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	/**
+	 * @param property
+	 *            the name of the property to get.
+	 * @param <T>
+	 *            the type of the property. Used for convenience.
+	 * 			
+	 * @return the property of this state with the given name, or
+	 *         <code>null</code> if it doesn't exist.
+	 */
+	@SuppressWarnings("unchecked")
```
Comment (Diocruel) [inline src/main/java/com/github/fishio/achievements/State.java:36]: Annoying. Though it is probably needed.
<layout

**Comment by Diocruel** [comment]
Discussed my comment with Jesse.

Good to merge
<feature,layout,organisation

**Pull request merged**